WeartGloves/src/WeArtGlove.cpp

```diff
@@ -0,0 +1,1598 @@
+
+        jointVelocity = Eigen::VectorXd::Zero(model.getNrOfDOFs());
+
+    }
+
+    return true;
+}
+
+bool WeArtGlove::WeArtGloveImpl::update()
+{
+    std::lock_guard<std::mutex> lock(mutex);
+
+    // To get the sensor raw data (IMUs and ToF)
+    for (auto &[key, element] : thimbleInformationMap)
+    {
+        int thimbleNumber = thimbleNameToNumber(key);
+
+        if (trackingAlgorithmType == "SCTA")
+        {
+            //Raw Closure data
+            weartClosure(thimbleNumber) = element.tracking->GetClosure();
+
+            // Apply exponential smoothing
+            weartClosureSmooth(thimbleNumber) = alpha * weartClosure(thimbleNumber) + (1 - alpha) * weartClosureSmooth(thimbleNumber);
+        }
+        else if (trackingAlgorithmType == "ASTA")
+        {
+            weartRawDataSample = element.rawSensorData->GetLastSample();
+
+            // Acceleration
+            weartRawDataAcc(thimbleNumber, 0) = weartRawDataSample.data.accelerometer.x;
+            weartRawDataAcc(thimbleNumber, 1) = weartRawDataSample.data.accelerometer.y;
+            weartRawDataAcc(thimbleNumber, 2) = weartRawDataSample.data.accelerometer.z;
+
+            // Gyroscope
+            weartRawDataGyro(thimbleNumber, 0) = weartRawDataSample.data.gyroscope.x;
+            weartRawDataGyro(thimbleNumber, 1) = weartRawDataSample.data.gyroscope.y;
+            weartRawDataGyro(thimbleNumber, 2) = weartRawDataSample.data.gyroscope.z;
+
+            // Raw ToF
+            weartRawDataTof(thimbleNumber) = weartRawDataSample.data.timeOfFlight.distance;
+        }
+
+        element.thimbleEffect->Set(element.thimbleEffect->effTemperature, element.thimbleEffect->effForce, element.thimbleEffect->effTexture);
+
+        if (element.hapticObject->activeEffects.size() <= 0)
+            element.hapticObject->AddEffect(element.thimbleEffect.get());
+        else
+            element.hapticObject->UpdateEffects();
+    }
+
+    if (trackingAlgorithmType == "SCTA")
+    {
+        humanJointState = closureToHumanJoint * weartClosureSmooth;
+
+        // saturate the computed values in the interval [0,1]
+        for (int i = 0; i < humanJointState.rows(); i++)
+        {
+            for (int j = 0; j < humanJointState.cols(); j++)
+            {
+                humanJointState(i, j) = std::max(0.0, std::min(1.0, humanJointState(i, j)));
+            }
+        }
+        humanJointState = humanJointState.cwiseProduct((MAX_JOINT_LIMIT - MIN_JOINT_LIMIT)) + MIN_JOINT_LIMIT;
+    }
+    else if (trackingAlgorithmType == "ASTA")
+    {
+        // Set the desired distance for the distance task
+        thumbDistanceTaskptr->setDesiredDistance(weartRawDataTof(0) / 1000.0);  // thumb
+        indexDistanceTaskptr->setDesiredDistance(weartRawDataTof(1) / 1000.0);  // index
+        middleDistanceTaskptr->setDesiredDistance(weartRawDataTof(2) / 1000.0); // middle
```

Comment:
Is the raw TOF in millimeters? Or 1000 is a scaling factor that works well? In general, it seems again a magic number

Reply:
It is in millimeters.

Reply:
Just make sure it is the case. In any case, consider adding a comment in the code